libs/metrics-wai/src/Data/Metrics/Types.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+module Data.Metrics.Types
+    ( PathTemplate(..)
+    , Paths(..)
+    , mkTree
+    , treeLookup
+    ) where
+
+import Imports
+import Data.Tree as Tree
+
+import qualified Data.ByteString.Char8 as BS
+
+
+newtype PathTemplate = PathTemplate Text
+
+-- | A 'Forest' of path segments.  A path segment is 'Nothing' if it captures a value
+-- (e.g. user id).
+newtype Paths = Paths (Forest (Maybe ByteString))
+  deriving (Eq, Show)
+
+-- | Turn a list of paths into a 'Paths' tree.  Treat all path segments that start with @':'@
+-- as equal and turn them into 'Nothing'.
+mkTree :: forall m. (m ~ Either String) => [[ByteString]] -> m Paths
```

Comment:
Hrmm, is there a particular reason you're using `m` here rather than just `Either String`?

Reply:
No reason, just a bad habit.  I wanted to make it a `MonadError` constraint (also out of bad habit), but then I decided against the extra dependency.

Reply:
Should I change it?

Reply:
I'd recommend the concrete type; the type errors for something like this are going to be much messier (I think?) since GHC will compare errors against `m` rather than `Either String` directly :)

Reply:
Error messages inside `mkTree` or on the call side?  When doing this change, I just found another (minor) reason why I like this weird idiom: once you have `m` declared as `Either String` in the signature of the top-level function, you can reuse it in other type declarations, which makes the code more concise and less redundant.

Anyway, I agree, kinda.  Changed!